src/data_structures/interval_tree.rs

```diff
@@ -0,0 +1,315 @@
+extern crate num;
+
+use self::num::traits::Num;
+
+use std::cmp;
+use std::mem;
+use std::fmt::Debug;
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct IntervalTree<N, D> {
+    root: Option<Node<N, D>>,
+}
+
+impl<N: Debug + Num + Clone + Ord, D: Debug> IntervalTree<N, D> {
+    pub fn new() -> Self {
+        IntervalTree { root: None }
+    }
+
+    pub fn insert(&mut self, start: N, end: N, data: D) {
+        validate_interval(&start, &end);
+        match self.root {
+            Some(ref mut n) => n.insert(start, end, data),
+            None => self.root = Some(Node::new(start, end, data)),
+        }
+    }
+
+    pub fn find_intersection(&self, start: N, end: N) -> Option<&D> {
+        validate_interval(&start, &end);
+        match self.root {
+            Some(ref n) => n.find_intersection(start, end),
+            None => None,
+        }
+    }
+}
+
+
+fn validate_interval<N: Debug + Num + Ord>(start: &N, end: &N) {
+    assert!(start <= end, "Invalid interval: [{:?}, {:?}]", start, end);
+}
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct Node<N, D> {
+    // actual interval data
+    start: N,
+    end: N,
```

Comment:
I also though about this regarding other parts of Rust-Bio recently: would you consider it a good idea to represent intervals as `Range<N>`? (a) this makes the semantic of `end` clear (inclusive or exclusive), (b) this allows intervals from here to be immediately used in slicing operations.